rust/index/src/fulltext/types.rs

```diff
@@ -86,55 +87,59 @@ impl<'me> FullTextIndexWriter<'me> {
 
     async fn populate_frequencies_and_posting_lists_from_previous_version(
         &self,
-        token: &str,
+        tokens: &Vec<Token>,
     ) -> Result<(), FullTextIndexError> {
-        let mut uncommitted_frequencies = self.uncommitted_frequencies.lock().await;
-        match uncommitted_frequencies.get(token) {
-            Some(_) => return Ok(()),
-            None => {
+        // (Scoped to limit the lifetime of the lock)
+        {
+            let mut uncommitted_frequencies = self.uncommitted_frequencies.lock().await;
+            for token in tokens {
+                if uncommitted_frequencies.contains_key(&token.text) {
+                    continue;
+                }
+
                 let frequency = match &self.full_text_index_reader {
                     // Readers are uninitialized until the first compaction finishes
                     // so there is a case when this is none hence not an error.
                     None => 0,
-                    Some(reader) => match reader.get_frequencies_for_token(token).await {
-                        Ok(frequency) => frequency,
-                        // New token so start with frequency of 0.
-                        Err(_) => 0,
-                    },
+                    Some(reader) => {
+                        match reader.get_frequencies_for_token(token.text.as_str()).await {
+                            Ok(frequency) => frequency,
+                            // New token so start with frequency of 0.
+                            Err(_) => 0,
+                        }
+                    }
                 };
                 uncommitted_frequencies
-                    .insert(token.to_string(), (frequency as i32, frequency as i32));
+                    .insert(token.text.clone(), (frequency as i32, frequency as i32));
             }
         }
+
         let mut uncommitted_postings = self.uncommitted_postings.lock().await;
-        match uncommitted_postings.positional_postings.get(token) {
-            Some(_) => {
-                // This should never happen -- if uncommitted has the token, then
-                // uncommitted_frequencies should have had it as well.
-                tracing::error!(
-                    "Error populating frequencies and posting lists from previous version"
-                );
-                return Err(FullTextIndexError::InvariantViolation);
```

Comment:
I can add an extra Vec to track seen tokens in the above loop so we can retain this error if it's important

Reply:
I think it is fine, not worth the extra overhead